### ISSUE TITLE
A Standard Schema for Accessibility Testing Results

### ISSUE THREAD

**Issue by bbertucc** [bbertucc/equalify#257]
## Overview
After experimenting with various scans and studying the work of @jrpool, I've started to see common ways to organize automated accessibility scan results. 

Developing a standard schema allows us to quickly develop new ways of understanding results from various scans.

Here is what I propose:

Our schema would organize findings into "messages," a term that accommodates a wide range of outputs, from errors and alerts to best practices and informational notes. 

A message represents a single finding or observation made by an accessibility tool. Messages are not limited to errors but can also include warnings, informational notes, and positive confirmations of accessibility best practices.

## Example Structure
```
{
  "messages": [
    {
      "name": "unique_identifier",
      "description": "Description of the message",
      "suggestion": "Suggestion for resolving the message",
      "detailsUrl": "URL to more information",
      "pages": [
          {
             "url": "url"
          }
       ],
      "tags": [
          {
             "name": "unique_identifier"
          }
       ],
      "elements": [
        "selector": "CSS selector of the element",
        "html": "Outer HTML of the affected element"
      },
      "relatedElements": [
        {
          "selector": "CSS selector of related element",
          "html": "Outer HTML of related element"
        }
      ]
    }
  ]
}
```

## Components
- **name**: A unique identifier for the rule or guideline that the message pertains to. This could be a rule ID from axe-core, a guideline reference from WAVE, or similar identifiers from other tools.
- **description**: A brief explanation of the issue, guideline, or best practice identified by the message.
- **suggestion**: Recommended actions or considerations for addressing the issue described in the message.
- **detailsUrl**: A direct link to more detailed information, guidelines, or remediation advice related to the message.
- **pages**: An array of pages related to the message. Many scans would output one-page url, but this should also support scans that output multiple pages.
- **tags**: An array of tags categorizing the message. Tags can include the name of the tool that generated the message (e.g., "axe-core", "WAVE"), the impact level of the issue (e.g., "critical"), relevant WCAG criteria (e.g., "WCAG2A"), and the nature of the message (e.g., "Error", "Best Practice").
- **element**: Information about the primary HTML element associated with the message, including a CSS selector for identification and the element's outer HTML for context.
- **relatedElements**: An array of objects representing additional HTML elements related to the message. Each object includes a CSS selector and the outer HTML of the related element.

## Incorporating Tool Outputs

**axe-core**
To integrate axe-core output into Equalify, map the id, description, impact (as a tag), and findings (nodes) from axe-core reports to the corresponding fields in the Equalify schema. axe-core's detailed node information, including selectors and HTML, aligns with the element and relatedElements components.

**WAVE**
WAVE tool outputs can be translated into Equalify messages by categorizing WAVE's findings (errors, alerts, features) as messages. The description and recommendations provided by WAVE fit the description and suggestion fields, respectively. Since WAVE categorizes its findings differently, use its categorizations as tags in the Equalify schema.

**Other Tools**
For other accessibility scanning tools, analyze the output format of each tool and map its findings to the Equalify schema. Key information to look for includes unique identifiers for issues, descriptions or names of the issues, impact levels, and any suggestions for remediation. HTML selectors and outer HTML provided by these tools can be directly used in the element and relatedElements fields.

**Comment by bbertucc** [comment]
@jrpool and @mgifford, tagging you here because I know you're both deep into automated testing. We're at the point with Equalify Version 2 developer where we're going to finalize our data architecture. Does this schema anticipate what you've seen with automated scan results?

**Comment by bbertucc** [comment]
I started a separate repo for this because I can see it being used for way more applications than just Equalify. Here it is: https://github.com/EqualifyEverything/equalify-formatter

**Comment by bbertucc** [comment]
EARL takes care of this: https://www.w3.org/TR/EARL10-Schema/

But we have to see if EARL is indeed viable for Equalify.

**Comment by bbertucc** [comment]
@jrpool- i'm kind of curious why you didn't just use EARL as a testaro standard?

**Comment by jrpool** [comment]
The outline above seems mainly to reflect what most tools output, with some qualifications:

1. One tool, Nu Html Validator, has no rule IDs. Testaro fabricates rule IDs for it in the form of regular expressions.
2. Some tools provide multi-paragraph suggestions.
3. If one message cites an array of pages and also an array of elements, then it would seem the elements are on various pages, but how would the page of each element be identified? In any case, existing tools almost always test only 1 page at any time.
4. The `tags` property informalizes much of what I believe a practical standardized reporting format should formalize. That, it seems to me, interferes with subsequent processing of reports.
5. The `element` property combines an unspecified set of facts that, in my opinion, merit individual formalization.
6. Tools use various methods for identifying the locations of elements, including line number, x-y coordinates, xpath, and CSS selector, not only CSS selector.
7. The outer HTML of elements can be bulky. Some defects belong to the HTML element, so its outer HTML can be megabytes in size.

**Comment by jrpool** [comment]
> @jrpool- i'm kind of curious why you didn't just use EARL as a testaro standard?

EARL seems to me to be a vocabulary rather than a schema. When used in the production of reports, it seems to be unnecessarily bulky. Its vocabulary for test results seems to be primitive, relying on plain language for presentation of details. It seems to be silent on the difficult decisions in reporting, such as how to classify severity and certainty, and how to identify elements and express relationships to standards. Empirical inspection of the outputs of the 10 tools integrated by Testaro has revealed similarities to one another that organically lead to a compromise reporting format such as the standardized Testaro standard result. When I then ask how I would re-express that with the EARL vocabulary, its documentation and examples make it difficult (at least for me) to answer that question, but they suggest the result would not be parsimonious. I would be happy to have these understandings corrected.

**Comment by bbertucc** [comment]
Totally agree @jrpool - It's a headache to understand! It could probably be better understood as JSON. The good news is the RDF format tat EARL uses lends itself to JSON pretty easily. For example, here is an EARL output:
```rdf
<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
         xmlns:earl="http://www.w3.org/ns/earl#">

  <earl:Assertion>
    <earl:assertedBy>
      <earl:Software>
        <rdf:type rdf:resource="http://www.example.com/AccessibilityChecker"/> 
      </earl:Software>
    </earl:assertedBy>
    <earl:subject>
      <rdf:Description rdf:about="http://www.example.com/productpage.html" /> 
    </earl:subject>
    <earl:test>
      <rdf:Description rdf:resource="http://www.w3.org/TR/WCAG20/#images-must-have-alt"/> 
    </earl:test>
    <earl:result>
      <earl:outcome earl:pass />
    </earl:result>
  </earl:Assertion>

</rdf:RDF>
```

Here it is in JSON:
```json
{
  "@context": {
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
    "earl": "http://www.w3.org/ns/earl#"
  },
  "@type": "earl:Assertion",
  "earl:assertedBy": {
    "@type": "earl:Software",
    "rdf:type": "http://www.example.com/AccessibilityChecker"
  },
  "earl:subject": {
    "@id": "http://www.example.com/productpage.html"
  },
  "earl:test": {
    "@id": "http://www.w3.org/TR/WCAG20/#images-must-have-alt"
  },
  "earl:result": {
    "earl:outcome": "earl:pass"
  }
}
```

Here's an even simpler JSON:
```
{
  "type": "Assertion",
  "assertedBy": {
    "type": "Software",
    "name": "AccessibilityChecker" 
  },
  "subject": "http://www.example.com/productpage.html",
  "test": "http://www.w3.org/TR/WCAG20/#images-must-have-alt",
  "result": "pass"
}
```

I'm wondering if we could work within that EARL format, using a simplified JSON schema?

**Comment by bbertucc** [comment]
Also might be worth mentioning that work on EARL stopped:
> At this time there are not sufficient implementations to complete EARL 1.0 Schema as a W3C Recommendation.

https://www.w3.org/TR/EARL10-Schema/

**Comment by jrpool** [comment]
As mentioned earlier, EARL seems to me to be a vocabulary, not a format, and its vocabulary seems to be silent on the important properties that accessibility testing tools report values of. It seems to abbreviate the entirety of test results into a single TestResult class, with no subordinate structure, so a description of a test result is nothing other than a string of text, such as “failed”.

So, the value of trying to use EARL as the basis, or as an inspiration, for a standard format is not yet clear to me.

**Comment by jrpool** [comment]
@bbertucc is correct that writing routines to convert tool outputs into a standard format requires nontrivial effort. In the Testaro project, that work is incorporated into the `standardize` module located in the `procs` directory. it converts the outputs of the 10 tools into a single format, created as a compromise reflecting the main properties that most tools report. When another tool is to be integrated, the `standardize` module is revised to cover the output of that tool, too.

**Comment by bbertucc** [comment]
Neat what you've done in `standardize` @jrpool! If we had a standard schema, you would be able to quickly integrate scan results without needing to standardize most things. The conversation is progressing [on A11y Slack](https://web-a11y.slack.com/archives/C04LZSN3T/p1709170562831039) with the folks behind axe-core and IBM's scanner. This also has a lot of implications for what the W3C team is doing with ACT standards: https://www.w3.org/WAI/standards-guidelines/act/report/testcases/

**Comment by bbertucc** [comment]
I'm moving forward with this issue here: https://github.com/EqualifyEverything/equalify-formatter

Once we have a standard format, I'm going to integrate it into Equalify.

**Comment by bbertucc** [comment]
[Equalify Schema](https://github.com/equalifyEverything/equalify-schema) is now live, so we can close this issue.
